File: language/dsl/fluent/src/core/base-builder/storage/value-storage.ts
import type {
  BaseBuildContext,
  FluentBuilder,
  MixedArrayMetadata,
} from "../types";
import { isFluentBuilder } from "../guards";

/**
 * Manages storage for builder property values
 *
 * Values are stored in three different maps based on their type:
 * - values: Static values (strings, numbers, plain objects without builders)
 * - builders: FluentBuilder instances and objects containing builders
 * - mixedArrays: Arrays containing both static values and builders
 *
 * This separation allows efficient resolution during the build process
 */
export class ValueStorage<T> {
  private values: Partial<T> = {};
  private builders: Map<
    string,
    FluentBuilder<unknown, BaseBuildContext> | Record<string, unknown>
  > = new Map();
  private mixedArrays: Map<string, MixedArrayMetadata> = new Map();

  constructor(initial?: Partial<T>) {
    if (initial) {
      this.values = { ...initial };
    }
  }

  /**
   * Sets a property value, intelligently routing it to the appropriate storage
   *
   * This method performs runtime type checking to determine how to store the value:
   * - FluentBuilder instances → builders Map
   * - Arrays with builders → mixedArrays Map
   * - Objects with builders → builders Map
   * - Everything else → values (static storage)
   */
  set<K extends keyof T>(key: K, value: unknown): void {
    const keyStr = String(key);

    if (isFluentBuilder(value)) {
      this.builders.set(keyStr, value);
      delete this.values[key];
      this.mixedArrays.delete(keyStr);
    } else if (Array.isArray(value)) {
      this.handleArrayValue(key, keyStr, value);
    } else if (
      typeof value === "object" &&
      value !== null &&
      this.containsBuilder(value)
    ) {
      this.builders.set(keyStr, value as Record<string, unknown>);
      delete this.values[key];
      this.mixedArrays.delete(keyStr);
    } else {
      this.setStaticValue(key, keyStr, value);
    }
  }

  /**
   * Handles array value storage, detecting mixed arrays with builders
   */
  private handleArrayValue<K extends keyof T>(
    key: K,
    keyStr: string,
    value: readonly unknown[],
  ): void {
    const builderIndices = new Set<number>();
    const objectIndices = new Set<number>();

    value.forEach((item, index) => {
      if (isFluentBuilder(item)) {
        builderIndices.add(index);
      } else if (
        typeof item === "object" &&
        item !== null &&
        this.containsBuilder(item)
      ) {
        objectIndices.add(index);
      }
    });

    const hasBuilders = builderIndices.size > 0 || objectIndices.size > 0;

    if (hasBuilders) {
      this.mixedArrays.set(keyStr, {
        array: value,
        builderIndices,
        objectIndices,
      });
      delete this.values[key];
    } else {
      this.setStaticValue(key, keyStr, value);
      this.mixedArrays.delete(keyStr);
    }

    this.builders.delete(keyStr);
  }

  /**
   * Sets a static value with proper type handling
   */
  private setStaticValue<K extends keyof T>(
    key: K,
    keyStr: string,
    value: unknown,
  ): void {
    // For known types in T, TypeScript will validate at compile time
    // At runtime, we trust the caller has provided a compatible type
    this.values[key] = value as T[K];
    this.builders.delete(keyStr);
    this.mixedArrays.delete(keyStr);
  }

  /**
   * Checks if an object contains any builders recursively
   * Handles circular references gracefully
   */
  private containsBuilder(
    obj: unknown,
    visited: WeakSet<object> = new WeakSet(),
  ): boolean {
    if (isFluentBuilder(obj)) return true;

    if (!obj || typeof obj !== "object") return false;

    if (visited.has(obj)) return false;
    visited.add(obj);

    if (Array.isArray(obj)) {
      return obj.some((item) => this.containsBuilder(item, visited));
    }

    const proto = Object.getPrototypeOf(obj);
    if (proto === Object.prototype || proto === null) {
      return Object.values(obj).some((val) =>
        this.containsBuilder(val, visited),
      );
    }

    return false;
  }

  /**
   * Checks if a property has been set
   */
  has<K extends keyof T>(key: K): boolean {
    const keyStr = String(key);
    return (
      key in this.values ||
      this.builders.has(keyStr) ||
      this.mixedArrays.has(keyStr)
    );
  }

  /**
   * Peeks at a property value without resolving builders
   * Returns the raw value if it's static, the array if it's mixed, or undefined if it's a builder
   */
  peek<K extends keyof T>(key: K): T[K] | undefined {
    const keyStr = String(key);

    const mixedArray = this.mixedArrays.get(keyStr);
    if (mixedArray) {
      return mixedArray.array as T[K];
    }

    if (this.builders.has(keyStr)) {
      return undefined;
    }

    return this.values[key];
  }

  /**
   * Gets builder for a property if one is set
   */
  peekBuilder<K extends keyof T, C extends BaseBuildContext>(
    key: K,
  ): FluentBuilder<T[K], C> | undefined {
    const keyStr = String(key);
    const entry = this.builders.get(keyStr);
    if (!entry) return undefined;

    if (isFluentBuilder<T[K], C>(entry)) {
      return entry;
    }

    return undefined;
  }

  /**
   * Gets the type of value stored for a property
   */
  getValueType<K extends keyof T>(
    key: K,
  ): "static" | "builder" | "mixed-array" | "unset" {
    const keyStr = String(key);

    if (this.mixedArrays.has(keyStr)) {
      return "mixed-array";
    }

    if (this.builders.has(keyStr)) {
      return "builder";
    }

    if (key in this.values) {
      return "static";
    }

    return "unset";
  }

  /**
   * Gets all static values for the build pipeline
   */
  getValues(): Readonly<Partial<T>> {
    return this.values;
  }

  /**
   * Gets all builder entries for the build pipeline
   */
  getBuilders(): ReadonlyMap<
    string,
    FluentBuilder<unknown, BaseBuildContext> | Record<string, unknown>
  > {
    return this.builders;
  }

  /**
   * Gets all mixed array entries for the build pipeline
   */
  getMixedArrays(): ReadonlyMap<string, MixedArrayMetadata> {
    return this.mixedArrays;
  }

  /**
   * Unsets a property, removing it from storage
   */
  unset<K extends keyof T>(key: K): void {
    const keyStr = String(key);
    delete this.values[key];
    this.builders.delete(keyStr);
    this.mixedArrays.delete(keyStr);
  }

  /**
   * Clears all properties from storage
   */
  clear(): void {
    this.values = {};
    this.builders.clear();
    this.mixedArrays.clear();
  }

  /**
   * Clones the storage, creating an independent copy
   */
  clone(): ValueStorage<T> {
    const cloned = new ValueStorage<T>();
    cloned.values = { ...this.values };
    cloned.builders = new Map(this.builders);
    cloned.mixedArrays = new Map(
      Array.from(this.mixedArrays.entries()).map(([key, metadata]) => [
        key,
        {
          array: metadata.array,
          builderIndices: new Set(metadata.builderIndices),
          objectIndices: new Set(metadata.objectIndices),
        },
      ]),
    );
    return cloned;
  }
}

File: language/dsl/kotlin/src/main/kotlin/com/intuit/playertools/fluent/core/ValueStorage.kt
package com.intuit.playertools.fluent.core

/**
 * Metadata for arrays that contain a mix of static values and builders.
 * Tracks which indices contain builders for selective resolution.
 */
data class MixedArrayMetadata(
    val array: List<Any?>,
    val builderIndices: Set<Int>,
    val objectIndices: Set<Int>,
)

/**
 * Determines the type of value stored for a given key.
 */
enum class ValueType {
    STATIC,
    BUILDER,
    MIXED_ARRAY,
    UNSET,
}

/**
 * Storage for builder property values with intelligent routing.
 * Separates static values, builder instances, and mixed arrays
 * for efficient resolution during the build phase.
 */
class ValueStorage {
    private val values = mutableMapOf<String, Any?>()
    private val builders = mutableMapOf<String, Any>()
    private val mixedArrays = mutableMapOf<String, MixedArrayMetadata>()

    /**
     * Sets a value with automatic routing based on type.
     * - FluentBuilder instances → builders map
     * - AssetWrapperBuilder instances → builders map
     * - Arrays containing builders → mixedArrays map
     * - Objects containing builders → builders map
     * - Everything else → values map
     */
    operator fun set(
        key: String,
        value: Any?,
    ) {
        when {
            value == null -> {
                values[key] = null
                builders.remove(key)
                mixedArrays.remove(key)
            }

            value is FluentBuilder<*> -> {
                builders[key] = value
                values.remove(key)
                mixedArrays.remove(key)
            }

            value is AssetWrapperBuilder -> {
                builders[key] = value
                values.remove(key)
                mixedArrays.remove(key)
            }

            value is List<*> -> {
                handleArrayValue(key, value)
            }

            value.containsBuilder() -> {
                builders[key] = value
                values.remove(key)
                mixedArrays.remove(key)
            }

            else -> {
                values[key] = value
                builders.remove(key)
                mixedArrays.remove(key)
            }
        }
    }

    /**
     * Gets a static value by key.
     */
    operator fun get(key: String): Any? = values[key]

    /**
     * Gets a builder by key.
     */
    fun getBuilder(key: String): Any? = builders[key]

    /**
     * Gets mixed array metadata by key.
     */
    fun getMixedArray(key: String): MixedArrayMetadata? = mixedArrays[key]

    /**
     * Checks if a key has any value (static, builder, or mixed array).
     */
    fun has(key: String): Boolean = key in values || key in builders || key in mixedArrays

    /**
     * Peeks at a value without resolution (checks all storage maps).
     */
    fun peek(key: String): Any? = values[key] ?: builders[key] ?: mixedArrays[key]?.array

    /**
     * Removes a value from all storage maps.
     */
    fun remove(key: String) {
        values.remove(key)
        builders.remove(key)
        mixedArrays.remove(key)
    }

    /**
     * Clears all stored values.
     */
    fun clear() {
        values.clear()
        builders.clear()
        mixedArrays.clear()
    }

    /**
     * Returns the type of value stored for a key.
     */
    fun getValueType(key: String): ValueType =
        when {
            key in mixedArrays -> ValueType.MIXED_ARRAY
            key in builders -> ValueType.BUILDER
            key in values -> ValueType.STATIC
            else -> ValueType.UNSET
        }

    /**
     * Creates a shallow clone of this storage.
     */
    fun clone(): ValueStorage =
        ValueStorage().also { cloned ->
            cloned.values.putAll(values)
            cloned.builders.putAll(builders)
            cloned.mixedArrays.putAll(mixedArrays)
        }

    /**
     * Returns all static values (for internal use in build pipeline).
     */
    internal fun getValues(): Map<String, Any?> = values.toMap()

    /**
     * Returns all builders (for internal use in build pipeline).
     */
    internal fun getBuilders(): Map<String, Any> = builders.toMap()

    /**
     * Returns all mixed arrays (for internal use in build pipeline).
     */
    internal fun getMixedArrays(): Map<String, MixedArrayMetadata> = mixedArrays.toMap()

    private fun handleArrayValue(
        key: String,
        value: List<*>,
    ) {
        val builderIndices = mutableSetOf<Int>()
        val objectIndices = mutableSetOf<Int>()

        value.forEachIndexed { index, item ->
            when {
                item is FluentBuilder<*> -> builderIndices.add(index)
                item is AssetWrapperBuilder -> builderIndices.add(index)
                item.containsBuilder() -> objectIndices.add(index)
            }
        }

        if (builderIndices.isNotEmpty() || objectIndices.isNotEmpty()) {
            mixedArrays[key] =
                MixedArrayMetadata(
                    array = value.toList(),
                    builderIndices = builderIndices,
                    objectIndices = objectIndices,
                )
            values.remove(key)
        } else {
            values[key] = value
            mixedArrays.remove(key)
        }
        builders.remove(key)
    }

    private fun Any?.containsBuilder(): Boolean {
        if (this == null) return false
        if (this is FluentBuilder<*>) return true
        if (this is AssetWrapperBuilder) return true
        if (this is Map<*, *>) return values.any { it.containsBuilder() }
        if (this is List<*>) return any { it.containsBuilder() }
        return false
    }
}
